src/WorkflowStep.ts

```diff
@@ -0,0 +1,379 @@
+    const { payload } = args;
+    const stepArgs = prepareStepArgs(args);
+    const stepMiddleware = this.getStepMiddleware(payload);
+    return processStepMiddleware(stepArgs, stepMiddleware);
+  }
+
+  private getStepMiddleware(payload: AllWorkflowStepMiddlewareArgs['payload']): WorkflowStepMiddleware {
+    switch (payload.type) {
+      case 'workflow_step_edit':
+        return this.edit;
+      case 'workflow_step':
+        return this.save;
+      case 'workflow_step_execute':
+        return this.execute;
+      default:
+        return [];
+    }
+  }
+}
+
+/** Helper Functions */
+
+export function validate(callbackId: string, config: WorkflowStepOptions): void {
+  // Ensure callbackId is valid
+  if (typeof callbackId !== 'string') {
+    const errorMsg = 'WorkflowStep expects a callback_id as the first argument';
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Ensure step config object is passed in
+  if (typeof config !== 'object') {
+    const errorMsg = 'WorkflowStep expects a configuration object as the second argument';
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Check for missing required keys
+  const requiredKeys: (keyof WorkflowStepOptions)[] = ['save', 'edit', 'execute'];
+  const missingKeys: (keyof WorkflowStepOptions)[] = [];
+  requiredKeys.forEach((key) => {
+    if (config[key] === undefined) {
+      missingKeys.push(key);
+    }
+  });
+
+  if (missingKeys.length > 0) {
+    const errorMsg = `WorkflowStep is missing required keys: ${missingKeys.join(', ')}`;
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Ensure a callback or an array of callbacks is present
+  const requiredFns: (keyof WorkflowStepOptions)[] = ['save', 'edit', 'execute'];
+  requiredFns.forEach((fn) => {
+    if (typeof config[fn] !== 'function' && !Array.isArray(config[fn])) {
+      const errorMsg = `WorkflowStep ${fn} property must be a function or an array of functions`;
+      throw new WorkflowStepInitializationError(errorMsg);
+    }
+  });
+}
+
+/**
+ * `processStepMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processStepMiddleware(
+  args: AllWorkflowStepMiddlewareArgs,
+  middleware: WorkflowStepMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  // TODO :: revisit type used below (look into contravariance)
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(callbacks, args, context, client, logger, async () =>
+      lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isStepEvent(args: AnyMiddlewareArgs): args is AllWorkflowStepMiddlewareArgs {
+  const validTypes = new Set(['workflow_step_edit', 'workflow_step', 'workflow_step_execute']);
```

Comment:
To avoid generating the same object every time running this function, how about instantiating `validTypes` outside this function (specifically, having `const` at the top level of this source and not having `export` for it should work for you)?